File: Data Structures & Algorithms/target-sum/submission-4.cpp
#include <vector>
#include <numeric>
using namespace std;

class Solution {
public:
    int findTargetSumWays(vector<int>& nums, int target) {
        int total = accumulate(nums.begin(), nums.end(), 0);

        if (abs(target) > total) return 0;
        if ((target + total) % 2) return 0;

        int P = (target + total) / 2;

        vector<int> dp(P + 1, 0);
        dp[0] = 1;

        for (int n : nums) {
            for (int s = P; s >= n; s--) {
                dp[s] += dp[s - n];
            }
        }

        return dp[P];
    }
};

File: Data Structures & Algorithms/squares-of-a-sorted-array/submission-4.py
class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        nums = [elem**2 for elem in nums]
        n = len(nums)
        res = [0]*n

        l = 0
        r = n-1
        m = n-1

        while l <= r:
            if nums[l] < nums[r]:
                res[m] = nums[r]
                r -= 1
                m -= 1
            else:
                res[m] = nums[l]
                l += 1
                m -= 1
        
        return res




        